main.cpp

```diff
@@ -225,33 +175,70 @@ float FloatType::divide( float lhs, float rhs )
 
 struct DoubleType
 {
-    double add( double lhs, double rhs );
-    double subtract( double lhs, double rhs );
-    double multiply( double lhs, double rhs );
-    double divide( double lhs, double rhs );   
+    double* valuePtr = new double;
+    double& value = *valuePtr;
+
+    DoubleType(double d) : value (d) {}
+    ~DoubleType()
+    {
+        delete valuePtr;
+        valuePtr = nullptr;
+    }
+
+    DoubleType& add(double rhs );
+    DoubleType& add(const FloatType f);
+
+    DoubleType& subtract(double rhs );
+    DoubleType& multiply(double rhs );
+    DoubleType& multiply(const IntType& i);
+    DoubleType& divide(double rhs );  
+    DoubleType& devide(const IntType& i); 
 };
 
 
-double DoubleType::add( double lhs, double rhs )
+DoubleType& DoubleType::add(double rhs )
+{
+    value += rhs;
+    return *this;
+}
+
+DoubleType& DoubleType::add(const FloatType i)
+{
+    value += i.value;
+    return *this;
+}
+
+
+DoubleType& DoubleType::subtract(double rhs )
 {
-    return lhs + rhs;
+    value -= rhs;
+    return *this;
 }
 
-double DoubleType::subtract( double lhs, double rhs )
+DoubleType& DoubleType::multiply(double rhs )
 {
-    return lhs - rhs;
+    value *= rhs;
+    return *this;
 }
 
-double DoubleType::multiply( double lhs, double rhs )
+DoubleType& DoubleType::multiply(const IntType& i)
 {
-    return lhs * rhs;
+    //value *= i.value;       // I know the IntType is only declared by name above (line 113). How can I give here the "complete" type of IntType?             
+    return *this;
 }
 
-double DoubleType::divide( double lhs, double rhs )
+DoubleType& DoubleType::divide(double rhs )
 {
     if(rhs == 0.0) 
         std::cout << "\nwarning, floating point division by zero returns 'inf' !" << std::endl;
-    return lhs / rhs;
+    value /= rhs;
+    return *this;
+}
+
+DoubleType& DoubleType::devide(const IntType& i)
+{
+    //value /= i.value;       // I know the IntType is only declared by name above (line 113). How can I give here the "complete" type of IntType? 
```

Comment:
You have to move this implementation to AFTER the definition of `IntType` is provided.
Code is interpreted by the compiler from the top-down, left to right.
Move all of the implementations to after each of the class declarations.
- [ ] revise